// ng-bootstrap-form-validation/esm5/ng-bootstrap-form-validation.js
import { CommonModule } from '@angular/common';
import { Component, ContentChildren, Directive, ElementRef, EventEmitter, HostBinding, HostListener, Inject, Injectable, InjectionToken, Input, NgModule, Output } from '@angular/core';
import { FormControlName, FormGroup } from '@angular/forms';
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var FormValidationDirective = /** @class */ (function () {
    function FormValidationDirective() {
        this.validSubmit = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FormValidationDirective.prototype.onSubmit = function () {
        this.markAsTouchedAndDirty(this.formGroup);
        if (this.formGroup.valid) {
            this.validSubmit.emit(this.formGroup.value);
        }
    };
    /**
     * @param {?} formGroup
     * @return {?}
     */
    FormValidationDirective.prototype.markAsTouchedAndDirty = function (formGroup) {
        var _this = this;
        Object.keys(formGroup.controls).forEach(function (key) {
            if (formGroup.controls[key] instanceof FormGroup) {
                _this.markAsTouchedAndDirty(/** @type {?} */ (formGroup.controls[key]));
            }
            else {
                formGroup.controls[key].markAsDirty();
                formGroup.controls[key].markAsTouched();
                formGroup.controls[key].updateValueAndValidity();
            }
        });
    };
    /**
     * @return {?}
     */
    FormValidationDirective.prototype.ngOnInit = function () { };
    return FormValidationDirective;
}());
FormValidationDirective.decorators = [
    { type: Directive, args: [{
                selector: "form[formGroup]"
            },] },
];
/** @nocollapse */
FormValidationDirective.ctorParameters = function () { return []; };
FormValidationDirective.propDecorators = {
    "formGroup": [{ type: Input },],
    "validSubmit": [{ type: Output },],
    "onSubmit": [{ type: HostListener, args: ["submit",] },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var DEFAULT_ERRORS = [
    {
        error: "required",
        format: function (label) { return label + " is required"; }
    },
    {
        error: "pattern",
        format: function (label) { return label + " is invalid"; }
    },
    {
        error: "minlength",
        format: function (label, error) { return label + " must be at least " + error.requiredLength + " characters"; }
    },
    {
        error: "maxlength",
        format: function (label, error) { return label + " must be no longer than " + error.requiredLength + " characters"; }
    },
    {
        error: "requiredTrue",
        format: function (label, error) { return label + " is required"; }
    },
    {
        error: "email",
        format: function (label, error) { return "Invalid email address"; }
    },
    {
        error: "max",
        format: function (label, error) { return label + " must be no greater than " + error.max; }
    },
    {
        error: "min",
        format: function (label, error) { return label + " must be no less than " + error.min; }
    }
];
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var CUSTOM_ERROR_MESSAGES = new InjectionToken("ng-bootstrap-form-validation custom error messages");
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var ErrorMessageService = /** @class */ (function () {
    /**
     * @param {?=} customErrorMessages
     */
    function ErrorMessageService(customErrorMessages) {
        if (customErrorMessages === void 0) { customErrorMessages = []; }
        this.customErrorMessages = customErrorMessages;
        this.defaultErrors = DEFAULT_ERRORS;
    }
    Object.defineProperty(ErrorMessageService.prototype, "errorMessages", {
        /**
         * @return {?}
         */
        get: function () {
            return this.customErrorMessages.concat(this.defaultErrors);
        },
        enumerable: true,
        configurable: true
    });
    return ErrorMessageService;
}());
ErrorMessageService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
ErrorMessageService.ctorParameters = function () { return [
    { type: Array, decorators: [{ type: Inject, args: [CUSTOM_ERROR_MESSAGES,] },] },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var FormGroupComponent = /** @class */ (function () {
    /**
     * @param {?} elRef
     * @param {?} errorMessageService
     */
    function FormGroupComponent(elRef, errorMessageService) {
        this.elRef = elRef;
        this.errorMessageService = errorMessageService;
        this.customErrorMessages = [];
        this.validationDisabled = false;
    }
    Object.defineProperty(FormGroupComponent.prototype, "hasErrors", {
        /**
         * @return {?}
         */
        get: function () {
            return (this.FormControlNames.some(function (c) { return !c.valid && c.dirty && c.touched; }) &&
                !this.validationDisabled);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormGroupComponent.prototype, "hasSuccess", {
        /**
         * @return {?}
         */
        get: function () {
            return (!this.FormControlNames.some(function (c) { return !c.valid; }) &&
                this.FormControlNames.some(function (c) { return c.dirty && c.touched; }) &&
                !this.validationDisabled);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormGroupComponent.prototype, "label", {
        /**
         * @return {?}
         */
        get: function () {
            var /** @type {?} */ label = this.elRef.nativeElement.querySelector("label");
            return label && label.textContent ? label.textContent.trim() : "This field";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormGroupComponent.prototype, "isDirtyAndTouched", {
        /**
         * @return {?}
         */
        get: function () {
            return this.FormControlNames.some(function (c) { return c.dirty && c.touched; });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormGroupComponent.prototype, "errorMessages", {
        /**
         * @return {?}
         */
        get: function () {
            return this.customErrorMessages.concat(this.errorMessageService.errorMessages);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormGroupComponent.prototype, "messages", {
        /**
         * @return {?}
         */
        get: function () {
            var _this = this;
            var /** @type {?} */ messages = [];
            if (!this.isDirtyAndTouched || this.validationDisabled)
                return messages;
            this.FormControlNames.filter(function (c) { return !c.valid; }).forEach(function (control) {
                Object.keys(control.errors).forEach(function (key) {
                    var /** @type {?} */ error = _this.errorMessages.find(function (error) { return error.error === key; });
                    if (!error)
                        return;
                    messages.push(error.format(_this.label, control.errors[key]));
                });
            });
            return messages;
        },
        enumerable: true,
        configurable: true
    });
    return FormGroupComponent;
}());
FormGroupComponent.decorators = [
    { type: Component, args: [{
                selector: ".form-group",
                template: "\n    <ng-content></ng-content>\n    <span class=\"help-block\" *ngFor=\"let message of messages\">{{message}}</span>\n  "
            },] },
];
/** @nocollapse */
FormGroupComponent.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: ErrorMessageService, },
]; };
FormGroupComponent.propDecorators = {
    "FormControlNames": [{ type: ContentChildren, args: [FormControlName,] },],
    "customErrorMessages": [{ type: Input },],
    "validationDisabled": [{ type: Input },],
    "hasErrors": [{ type: HostBinding, args: ["class.has-error",] },],
    "hasSuccess": [{ type: HostBinding, args: ["class.has-success",] },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgBootstrapFormValidationModule = /** @class */ (function () {
    function NgBootstrapFormValidationModule() {
    }
    /**
     * @param {?=} customErrorMessages
     * @return {?}
     */
    NgBootstrapFormValidationModule.forRoot = function (customErrorMessages) {
        return {
            ngModule: NgBootstrapFormValidationModule,
            providers: [
                {
                    provide: ErrorMessageService,
                    useFactory: errorMessageServiceFactory,
                    deps: [CUSTOM_ERROR_MESSAGES]
                },
                {
                    provide: CUSTOM_ERROR_MESSAGES,
                    useValue: customErrorMessages
                }
            ]
        };
    };
    return NgBootstrapFormValidationModule;
}());
NgBootstrapFormValidationModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormValidationDirective, FormGroupComponent],
                imports: [CommonModule],
                providers: [ErrorMessageService],
                exports: [FormValidationDirective, FormGroupComponent]
            },] },
];
/** @nocollapse */
NgBootstrapFormValidationModule.ctorParameters = function () { return []; };
/**
 * @param {?=} customErrorMessages
 * @return {?}
 */
function errorMessageServiceFactory(customErrorMessages) {
    return new ErrorMessageService(customErrorMessages);
}
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { errorMessageServiceFactory, NgBootstrapFormValidationModule, FormValidationDirective, FormGroupComponent, ErrorMessageService as ɵa, CUSTOM_ERROR_MESSAGES as ɵb };
